### PR TITLE
[NativeAOT-LLVM] Add a native WASM object writer

### PR DESCRIPTION
**Pull request by SingleAccretion**
This adds an object writer that directly emits WASM object files (see https://github.com/WebAssembly/tool-conventions/blob/main/Linking.md) and uses it to emit data symbols. It is anticipated that future changes will use it for code (compiler-generated helpers) as well.

We have also decided to delete the LLVM-based object writer, since doing otherwise would imply either maintaining it (duplication of effort) or letting it bitrot. It means that ILC will no longer produce all output as LLVM bitcode.

There are two motivations for this change:
1) (Primary) https://github.com/dotnet/runtimelab/issues/2404 - we need to be able to emit relocations LLVM doesn't support.
2) (Secondary) Performance: the new object writer will, of course, skip all of the LLVM steps, which have been observed to be rather lengthy on larger projects due to some algorithmic issues in LLVM.

### CODE REVIEW

**Comment by yowl** [comment]
This looks great!

**Comment by SingleAccretion** [comment]
@dotnet/nativeaot-llvm

File: src/coreclr/tools/aot/ILCompiler.LLVM/CodeGen/LLVMCodegenCompilation.CodeGen.cs [hunk 1]
```diff
@@ -47,6 +48,116 @@ public override ISymbolNode GetExternalMethodAccessor(MethodDesc method, ReadOnl
 
         public override CorInfoLlvmEHModel GetLlvmExceptionHandlingModel() => Options.ExceptionHandlingModel;
 
+        internal WasmFunctionType GetWasmFunctionTypeForMethod(MethodSignature signature, bool hasHiddenParam)
+        {
+            WasmValueType wasmPointerType = _nodeFactory.Target.PointerSize == 4 ? WasmValueType.I32 : WasmValueType.I64;
+
+            WasmValueType GetWasmTypeForTypeDesc(TypeDesc type)
```
Comment (kg) [inline src/coreclr/tools/aot/ILCompiler.LLVM/CodeGen/LLVMCodegenCompilation.CodeGen.cs:55]: Just curious, has struct scalarization (from the WASM C ABI) been performed at this point? Or do you have to do it?

Comment (SingleAccretion) [reply src/coreclr/tools/aot/ILCompiler.LLVM/CodeGen/LLVMCodegenCompilation.CodeGen.cs:55]: Yes, it has - all of the callers of this function first check `GetPrimitiveTypeForTrivialWasmStruct`. So here we only expect primitives.

Comment (yowl) [reply src/coreclr/tools/aot/ILCompiler.LLVM/CodeGen/LLVMCodegenCompilation.CodeGen.cs:55]: Is this related to multi value return types or does LLVM never produce those?

Comment (SingleAccretion) [reply src/coreclr/tools/aot/ILCompiler.LLVM/CodeGen/LLVMCodegenCompilation.CodeGen.cs:55]: We don't support the multi-value calling convention. It is not standardized yet either.

Comment (yowl) [reply src/coreclr/tools/aot/ILCompiler.LLVM/CodeGen/LLVMCodegenCompilation.CodeGen.cs:55]: https://webassembly.org/features/  suggests it is standardized, but if we don't support it then no problem.

**Review by yowl — APPROVED** [review]
This looks good to me.  I'm not familiar with the object file layout, so I'm going by the fact that as it works, it is correct.  Do we have any before/after numbers for total wasm size ?

**Comment by SingleAccretion** [comment]
> Do we have any before/after numbers for total wasm size ?

Yes - this is a no-diff change, as expected.

**Comment by jkotas** [comment]
Thank you!

**Pull request merged**